src/common/src/cache.rs

```diff
@@ -769,65 +785,102 @@ impl<K: LruKey, T: LruValue> LruCache<K, T> {
 
 pub struct CleanCacheGuard<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> {
     cache: &'a Arc<LruCache<K, T>>,
-    key: K,
+    key: Option<K>,
     hash: u64,
-    success: bool,
+}
+
+impl<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> CleanCacheGuard<'a, K, T> {
+    fn mark_success(mut self) -> K {
+        self.key.take().unwrap()
+    }
 }
 
 impl<'a, K: LruKey + Clone + 'static, T: LruValue + 'static> Drop for CleanCacheGuard<'a, K, T> {
     fn drop(&mut self) {
-        if !self.success {
-            self.cache.clear_pending_request(&self.key, self.hash);
+        if let Some(key) = self.key.as_ref() {
+            self.cache.clear_pending_request(key, self.hash);
+        }
+    }
+}
+
+/// `lookup_with_request_dedup` can directly return `Result<CacheableEntry<K, T>, E>`, but if we do
+/// not want to wait when cache hit does not happen, we can call `lookup_with_request_dedup_noawait`
+/// which will return `BoxFuture<'static, Result<CacheableEntry<K, T>, E>>` when cache hit does not
+/// happen.
+pub enum LookupResponse<K: LruKey + Clone + 'static, T: LruValue + 'static, E> {
+    Cached(CacheableEntry<K, T>),
+    WaitPendingRequest(Receiver<CacheableEntry<K, T>>),
+    Miss(JoinHandle<Result<CacheableEntry<K, T>, E>>),
+}
+
+impl<K: LruKey + Clone + 'static, T: LruValue + 'static, E: From<RecvError>>
```

Comment:
> LookupResponse

I read code in `UploadHandleManagerNextFinishedEpoch` but still do not understand how to store `Waker` in `LookupResponse`. Is there any example which shows that how to store `Waker`?

Reply:
> Can we `impl Future` for the `LookupResponse`, so that we don't have to separated `lookup_with_request_dedup_noawait` and `lookup_with_request_dedup`? We can have a single `fn lookup_with_request_dedup(...)` which returns the `LookupResponse`. The current `lookup_with_request_dedup(...).await` still works the same.

I read code in `UploadHandleManagerNextFinishedEpoch` but still do not understand how to store `Waker` in `LookupResponse`. Is there any example which shows that how to store `Waker`?

Reply:
We don't have to store the `Waker`. 

Code will be like 
```
impl Future for LookupResponse {
  Output = Result<Entry<...>>
  fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
    match &mut self {
        Cached(entry) => Poll::Ready(entry),
        Pending(recv) => recv.poll(cx),
    }
  }
}
```

In practice, awaken the current tokio task is the work of the underlying `recv` which blocks the current task, and we don't have to store the waker in our code. The waker is stored in the `recv` when we call `recv.poll(cx)` on it and will awake the current task when the work of `recv` finishes(the sender sends something and wake).

Reply:
It seems that the requirement on `HummockIterator` to be `Sync` has blocked our attempt to use `BoxedFuture`. I have tried removing the `Sync` requirement on the `HummockIterator` trait definition, and the code compiles without any error. So maybe we can try simply removing the `Sync` requirement on the `HummockIterator` definition, and things may get easier.

Reply:
> current tokio task is the work of the underlying `recv` which blocks the current task, and we don't have to store

thanks. However it seems that `mpsc::receiver` has method `poll_recv` but `mpsc::oneshot` does not.

Reply:
> Since `oneshot::receiver` has `impl Future`, it has `poll` already. Since it is also `Unpin`, we can directly call `poll_unpin` on it.
> 
> If the implementation is complex, we can simply return a `BoxedFuture` and remove the `Sync` requirement on the `trait HummockIterator`.

done